diff --git a/advanced_forum.module b/advanced_forum.module
index 82f4df8..65092fd 100644
--- a/advanced_forum.module
+++ b/advanced_forum.module
@@ -880,7 +880,11 @@ function advanced_forum_get_reply_link($node) {
     }

     if ($allowed) {
-      $fragment = $node->content['comments']['comment_form']['#id'];
+      if (!isset($node->content)) {
+        $node->content['comments'] = comment_node_page_additions($node);
+      }
+      $fragment = isset($node->content['comments']['comment_form']['#id']) ?
+        $node->content['comments']['comment_form']['#id'] : "comment-form";

       if (variable_get('comment_form_location_' . $node->type, COMMENT_FORM_SEPARATE_PAGE) == COMMENT_FORM_SEPARATE_PAGE) {
         // Reply form is on separate page. Grab the href from the node links
