diff --git a/ip2country.inc b/ip2country.inc
index 59ac5f2..9e1b0b5 100644
--- a/ip2country.inc
+++ b/ip2country.inc
@@ -215,18 +215,7 @@ function ip2country_update_database($registry = 'arin') {
  *   of the fetched file.
  */
 function _ip2country_fetch_page($url) {
-  $curl = curl_init();
-
-  // Fetch requested file.
-  curl_setopt($curl, CURLOPT_URL, $url);
-  curl_setopt($curl, CURLOPT_TIMEOUT, 60*2);
-  curl_setopt($curl, CURLOPT_USERAGENT, 'Drupal (+http://drupal.org/)');
-  curl_setopt($curl, CURLOPT_HEADER, FALSE);
-  curl_setopt($curl, CURLOPT_RETURNTRANSFER, TRUE);
-  curl_setopt($curl, CURLOPT_FOLLOWLOCATION, TRUE);
-
-  $html = curl_exec($curl);
-  curl_close($curl);
+  $html = file_get_contents($url);
 
   return $html;
 }