diff --git a/views_autocomplete_filters.inc b/views_autocomplete_filters.inc
index c655cb4..c785283 100644
--- a/views_autocomplete_filters.inc
+++ b/views_autocomplete_filters.inc
@@ -46,7 +46,7 @@ function views_autocomplete_filter($filter_name, $view_name, $display_name, $vie
   $expose_options = $filter['expose'];
 
   // Do not filter if the string length is less that minimum characters setting.
-  if (drupal_strlen(trim($string)) < $expose_options['autocomplete_min_chars']) {
+  if (isset($expose_options['autocomplete_min_chars']) && drupal_strlen(trim($string)) < $expose_options['autocomplete_min_chars']) {
     $matches[''] = '<div class="reference-autocomplete">' . t('Search text should have at least %min_chars characters.', array('%min_chars' => $expose_options['autocomplete_min_chars'])) . '</div>';
     return drupal_json_output($matches);
   }