Index: sites/all/modules/contrib/wysiwyg/editors/js/ckeditor-3.0.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sites/all/modules/contrib/wysiwyg/editors/js/ckeditor-3.0.js	(date 1496326422000)
+++ sites/all/modules/contrib/wysiwyg/editors/js/ckeditor-3.0.js	(revision )
@@ -20,7 +20,6 @@
  *   An object containing global plugin configuration.
  */
 Drupal.wysiwyg.editor.init.ckeditor = function(settings, pluginInfo) {
-  instanceMap = {};
   // Nothing to do here other than register new plugins etc.
   Drupal.wysiwyg.editor.update.ckeditor(settings, pluginInfo);
 };
@@ -182,6 +181,9 @@
       delete instanceMap[this.name];
     }
   };
+  if ("undefined" === typeof instanceMap) {
+      instanceMap = {};
+  }
   instanceMap[params.field] = this;
   // Attach editor.
   var editorInstance = CKEDITOR.replace(params.field, settings);